src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs

```diff
@@ -532,7 +532,15 @@ public void IntersectWith(IEnumerable<T> other)
                 // that other is a hashset using the same equality comparer.
                 if (other is HashSet<T> otherAsSet && EqualityComparersAreEqual(this, otherAsSet))
                 {
-                    IntersectWithHashSetWithSameComparer(otherAsSet);
+                    IntersectWithCollection(otherAsCollection);
+                    return;
+                }
+
+                Type? otherType = other.GetType().GetGenericTypeDefinition();
+                // For types supporting O(1) lookup, utilize the contains method for faster intersect.
+                if (otherType == typeof(Dictionary<,>.KeyCollection))
```

Comment:
Does this need to verify that the comparer is the same?

This may need new functional tests to validate that this combination is detected and works correctly.

Reply:
> Does this need to verify that the comparer is the same?

The existing implementation which treats it as a `IEnumerable` doesn't seem to verify the comparer.

Ref: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs#L1295

Reply:
The behavior of `IEnumerable` does not depend on comparer.

Reply:
Right. I need to re - look this one with test cases. Thanks.

Reply:
@jkotas For dictionary, `KeyCollection` is being passed and not the dictionary itself. In this case we may never come know what comparer was used in the original dictionary.

Reply:
It suggests that it is not possible to implement this optimization correctly.

Reply:
Didn't get this one. Doesn't the current implementation have the same behavior where the comparer gets ignored ?

Reply:
The current implementation is fine. It is enumerating everything for this case. Enumeration does not depend on comparer. It always enumerates everything.

Reply:
So is it by concept that it would be wrong to implement such a change ? In the current implementation, it simply iterates over each element and checks whether it is in `HashSet` and in this implementation, `Dictionary.KeyCollection`'s contains is invoked which also does the same.

Because by behavior, i.e. calling intersect with a dictionary's key collection yields the same results in both cases. I will add test cases for this.

Reply:
Here is a program that demonstrates the problem:
```
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;

class LowBitsEqualityComparer : IEqualityComparer<int>
{
    public bool Equals(int x, int y) => (x & 15) == (y & 15);
    public int GetHashCode(int obj) => obj & 15;
}

class Program
{
    static void Main(string[] args)
    {
        var hashSet = new HashSet<int>() { 1, 17 };
        var dictionary = new Dictionary<int, bool>(new LowBitsEqualityComparer()) { { 33, true } };

        hashSet.IntersectWith(dictionary.Keys);

        Console.WriteLine(hashSet.Count);
    }
}
```

It prints 0 before this change and 2 after this change.

Reply:
Ah. Got it now. Thanks.

Reply:
@jkotas Sorry to bother again. I was playing with the above sample and tried a few more operations. Consider the following: 
```
using System;
using System.Linq;
using System.Collections.Generic;
				
class LowBitsEqualityComparer : IEqualityComparer<int>
{
    public bool Equals(int x, int y) => (x & 15) == (y & 15);
    public int GetHashCode(int obj) => obj & 15;
}

public class Program
{
	public static void Main()
	{
		var hashSet = new HashSet<int>() { 1, 17 };
        var dictionary = new Dictionary<int, bool>(new LowBitsEqualityComparer()) { { 33, true } };

		var dictionaryKeyIntersect = dictionary.Keys.Intersect(hashSet);
        hashSet.IntersectWith(dictionary.Keys);

		// 0
		Console.WriteLine(dictionaryKeyIntersect.Count());
		// 0
        Console.WriteLine(hashSet.Count);
		
		// True
		Console.WriteLine(dictionary.Keys.Contains(1));
		// True
		Console.WriteLine(dictionary.Keys.Contains(17));
	}
}
```
 Now, here I know that the above behaviour is happening because of `IEnumerable` comparison with default comparer happening in the intersect case and  in the case of  calling `Contains` on `Dictionary.KeyCollection`, it calls the `Dictionary`'s `ContainsKey` method which uses the `IComparer` while comparing elements.

What I wanted to know was whether is this an ideal behaviour that should happen ? Because as a user it is a bit confusing when calling intersect from either side, I get no elements but `Contains` is returning true.

Reply:
If a single function uses collections with same key and different equality comparer, I agree that it is easy to get lost in it and end up with a code that has confusing behavior. I think the solution to that is to comment the code well and write it such that it is clear what is going on.